packages/ic-hpke/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+        // under a different version
+        //
+        // An alternative, had the info parameter not been available, would be to
+        // concatenate the user-provided associated data and the header
+        // together. However this separation is neater and avoids having to
+        // allocate and copy the associated data.
+
+        let (hpke_key, hpke_ctext) = hpke::single_shot_seal::<V1Aead, V1Kdf, V1Kem, R>(
+            opmode,
+            &self.pk,
+            &buf,
+            msg,
+            associated_data,
+            rng,
+        )
+        .map_err(|_| EncryptionError::InternalError)?;
+
+        buf.extend_from_slice(&hpke_key.to_bytes());
+        buf.extend_from_slice(&hpke_ctext);
+
+        Ok(buf)
+    }
+
+    /// Serialize the public key to a bytestring
+    pub fn serialize(&self) -> Vec<u8> {
+        let len = <V1PublicKey as Serializable>::size();
+        let mut buf = vec![0u8; HEADER_SIZE + len];
+        buf[0..HEADER_SIZE].copy_from_slice(&MAGIC.to_be_bytes());
+        self.pk.write_exact(&mut buf[HEADER_SIZE..]);
+        buf
+    }
+
+    /// Deserialize a public key
+    pub fn deserialize(bytes: &[u8]) -> Result<Self, KeyDeserializationError> {
+        if bytes.len() < HEADER_SIZE {
+            return Err(KeyDeserializationError::InvalidLength);
+        }
+
+        let magic = u64::from_be_bytes(
+            bytes[0..HEADER_SIZE]
+                .try_into()
+                .expect("Conversion cannot fail"),
+        );
+        if magic != MAGIC {
+            return Err(KeyDeserializationError::UnknownMagic);
+        }
+
+        let len = <V1PublicKey as Serializable>::size();
+        if bytes.len() != HEADER_SIZE + len {
+            return Err(KeyDeserializationError::InvalidLength);
+        }
+
+        match V1PublicKey::from_bytes(&bytes[HEADER_SIZE..]) {
+            Ok(pk) => Ok(Self { pk }),
+            Err(_) => Err(KeyDeserializationError::InvalidKey),
+        }
+    }
+
+    fn new(pk: V1PublicKey) -> Self {
+        Self { pk }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+/// An error occured while decrypting a message
+pub enum DecryptionError {
+    /// The protocol identifier/version field did not match
+    UnknownMagic,
+    /// The length was wrong
+    InvalidLength,
+    /// The header was valid but the decryption failed
+    InvalidCiphertext,
+}
+
+#[derive(Clone)]
+/// A private key usable for decryption
+pub struct PrivateKey {
+    sk: V1PrivateKey,
+    pk: V1PublicKey,
+}
+
+impl PrivateKey {
+    /// Generate a new random private key
+    pub fn generate<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let mut ikm = vec![0; 64];
+        rng.fill_bytes(&mut ikm);
+        let (sk, pk) = <V1Kem as Kem>::derive_keypair(&ikm);
+        Self { sk, pk }
+    }
+
+    /// Decrypt a message without sender authentication
+    ///
+    /// This is the counterpart to [`PublicKey::encrypt_noauth`]
```

Comment:
It may be worth to explicitly mention that this cannot be combined with `PublicKey::encrypt_noauth`.

Reply:
I think you meant here "cannot be combined with `PublicKey::encrypt`" (and yes good idea to point this out, as otherwise this aspect is likely to be confusing)

Reply:
Mentioned this in the doc comment

Reply:
> I think you meant here "cannot be combined with PublicKey::encrypt"

Yes, indeed (had a copy/paste error in my sentence)!

Thanks for updating the docu.